l2l/optimizers/kalmanfilter/enkf.py

```diff
@@ -0,0 +1,155 @@
+
+        :param maxit: int, maximum number of iterations
+        :param n_batches, int,  number of batches to used in mini-batch. If set
+            to `1` uses the whole given dataset. Default is `1`.
+        :param online, bool, True if one random data point is requested,
+            between [0, dims], otherwise do mini-batch. `dims` is the number of
+            observations. Default is False
+        """
+        self.Cpp = None
+        self.Cup = None
+        self.ensemble = None
+        self.observations = None
+
+        self.maxit = maxit
+        self.online = online
+        self.n_batches = n_batches
+        self.gamma = 0.
+        self.gamma_s = 0
+        self.dims = 0
+
+    def fit(self, ensemble, ensemble_size,
+            observations, model_output, gamma, noise=0., p=None, u_exact=None):
+        """
+        Prediction and update step of the EnKF
+
+        Calculates new ensembles.
+
+        :param ensemble: nd numpy array, contains ensembles `u`
+        :param ensemble_size: int, number of ensembles
+        :param u_exact: nd numpy array, exact control, e.g. if weight sof the model
+                   are known. Default is `None`
+        :param observations: nd numpy array, observation or targets
+        :param model_output: nd numpy array, output of the model
+            In terms of the Kalman Filter the model maps the ensembles (dim n)
+            into the observed data `y` (dim k). E.g. network output activity
+        :param noise: nd numpy array, Noise can be added to the model (for `gamma`)
+            and is used in the misfit calculation for convergence.
+            E.g. multivariate normal distribution. Default is `0.0`
+        :param  gamma: nd numpy array, Normalizes the model-data distance in the
+            update step, :`noise * I` (I is identity matrix) or
+            :math:`\\gamma=I` if `noise` is zero
+        :param p: nd numpy array
+            Exact solution given by :math:`G * u_exact`, where `G` is inverse
+            of a linear elliptic function `L`, it maps the control into the
+            observed data, Default is `None`
+        :return self, Possible outputs are:
+            ensembles: nd numpy array, optimized `ensembles`
+            Cpp: nd numpy array, covariance matrix of the model output
+            Cup: nd numpy array, covariance matrix of the model output and the
+                ensembles
+        """
+        model_output = model_output.copy()
+        # get shapes
+        self.gamma_s, self.dims = _get_shapes(observations, model_output)
+
+        if isinstance(gamma, (int, float)):
+            if float(gamma) == 0.:
+                self.gamma = np.eye(self.gamma_s)
+        else:
+            self.gamma = gamma
+
+        # copy the data so we do not overwrite the original arguments
+        self.ensemble = ensemble.copy()
+        self.observations = observations.copy()
+        self.observations = _encode_targets(observations, self.gamma_s)
```

Comment:
It looks like you might be overwriting `self.observations` with this value calculated from `observations`. So the previous line has no effect.